File: gaia/frontend/src/lib/voice.ts
/**
 * Typed client for GAIA's voice endpoints.
 *
 * Deliberately outside `stream.ts`/the chat SSE contract: STT runs before a
 * normal `send()` call (the transcribed text is indistinguishable from
 * typing once it reaches the chat store), and TTS runs after one, on the
 * already-persisted final assistant text. Neither call is part of a chat
 * turn — see docs/ARCHITECTURE.md, "Voice".
 */

import { apiBase, ApiError } from './api'

export interface TranscribeResult {
  text: string
  language: string | null
}

async function readErrorDetail(response: Response, fallback: string): Promise<string> {
  try {
    const body = await response.json()
    if (typeof body?.detail === 'string') return body.detail
    if (Array.isArray(body?.detail)) return body.detail[0]?.msg ?? fallback
  } catch {
    /* keep the fallback */
  }
  return fallback
}

export const voiceApi = {
  /** Upload a recorded clip, get back the transcribed text. */
  async transcribe(blob: Blob): Promise<TranscribeResult> {
    const form = new FormData()
    form.append('file', blob, 'recording.audio')

    let response: Response
    try {
      response = await fetch(`${apiBase()}/api/voice/transcribe`, { method: 'POST', body: form })
    } catch {
      throw new ApiError('Could not reach the GAIA backend. Is it running?', 0)
    }
    if (!response.ok) {
      throw new ApiError(
        await readErrorDetail(response, `Transcription failed (${response.status}).`),
        response.status,
      )
    }
    return (await response.json()) as TranscribeResult
  },

  /** Synthesize `text` to speech, get back playable audio. */
  async speak(text: string): Promise<Blob> {
    let response: Response
    try {
      response = await fetch(`${apiBase()}/api/voice/speak`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ text }),
      })
    } catch {
      throw new ApiError('Could not reach the GAIA backend. Is it running?', 0)
    }
    if (!response.ok) {
      throw new ApiError(
        await readErrorDetail(response, `Speech synthesis failed (${response.status}).`),
        response.status,
      )
    }
    return await response.blob()
  },
}

File: gaia/backend/gaia/api/voice.py
"""Voice endpoints — HTTP shape only, no engine logic here.

Deliberately outside the chat turn: `/transcribe` runs *before* a normal
`POST /api/chat` call (the transcribed text becomes an ordinary `content`
string — the chat pipeline cannot tell it apart from typing), and `/speak`
runs *after* one, on the already-persisted final assistant text. Neither
endpoint touches `chat_service.py`, `ToolCall`, or the SSE stream.

Every temporary audio file this module creates — incoming recordings and
synthesized speech alike — is deleted in a `finally` block before the
response is returned. Nothing here is meant to outlive its own request; see
`config.voice_dir`'s docstring and docs/PRIVACY.md, "Voice".
"""

from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, HTTPException, Response, UploadFile
from sqlalchemy.orm import Session

from gaia.config import get_settings
from gaia.db.session import get_session
from gaia.schemas.api import SpeakRequest, TranscribeResponse
from gaia.services import settings_service
from gaia.voice.base import VoiceProviderError
from gaia.voice.registry import build_stt_provider, build_tts_provider

router = APIRouter(prefix="/api/voice", tags=["voice"])

#: Generous for a few minutes of push-to-talk audio; nowhere near unbounded.
_MAX_AUDIO_BYTES = 25 * 1024 * 1024


@router.post("/transcribe", response_model=TranscribeResponse)
async def transcribe(
    file: UploadFile, session: Session = Depends(get_session)
) -> TranscribeResponse:
    payload = await file.read()
    if not payload:
        raise HTTPException(status_code=422, detail="The uploaded recording is empty.")
    if len(payload) > _MAX_AUDIO_BYTES:
        raise HTTPException(status_code=413, detail="Recording is too large.")

    settings = get_settings()
    settings.ensure_directories()
    # The extension doesn't need to match the browser's real codec exactly —
    # ffmpeg (via `av`, faster-whisper's decode path) sniffs the container
    # from its contents, not the filename.
    temp_path = settings.voice_dir / f"stt-{uuid.uuid4().hex}.audio"
    temp_path.write_bytes(payload)
    try:
        provider_id = settings_service.get(session, settings_service.STT_PROVIDER)
        provider = build_stt_provider(provider_id)
        result = await provider.transcribe(str(temp_path))
    except VoiceProviderError as exc:
        raise HTTPException(status_code=502, detail=exc.message) from exc
    finally:
        # The raw recording is never kept beyond the single request that needed it.
        temp_path.unlink(missing_ok=True)

    if not result.text.strip():
        raise HTTPException(
            status_code=422, detail="Could not make out any speech in that recording."
        )
    return TranscribeResponse(text=result.text, language=result.language)


@router.post("/speak")
async def speak(payload: SpeakRequest, session: Session = Depends(get_session)) -> Response:
    settings = get_settings()
    settings.ensure_directories()
    out_path = settings.voice_dir / f"tts-{uuid.uuid4().hex}.wav"
    try:
        provider_id = settings_service.get(session, settings_service.TTS_PROVIDER)
        provider = build_tts_provider(provider_id)
        await provider.synthesize(payload.text, out_path=str(out_path))
        audio_bytes = out_path.read_bytes()
    except VoiceProviderError as exc:
        raise HTTPException(status_code=502, detail=exc.message) from exc
    finally:
        # The synthesized clip is read into the response body above and then
        # deleted here — nothing is kept on disk after this request returns.
        out_path.unlink(missing_ok=True)

    return Response(content=audio_bytes, media_type="audio/wav")
